src/cargo/ops/common_for_install_and_uninstall.rs

```diff
@@ -1,45 +1,481 @@
+        } else {
+            None
+        };
+
+        let v1 = (|| -> CargoResult<_> {
+            let mut contents = String::new();
+            v1_lock.file().read_to_string(&mut contents)?;
+            if contents.is_empty() {
+                Ok(CrateListingV1::default())
+            } else {
+                Ok(toml::from_str(&contents)
+                    .chain_err(|| format_err!("invalid TOML found for metadata"))?)
+            }
+        })()
+        .chain_err(|| {
+            format_err!(
+                "failed to parse crate metadata at `{}`",
+                v1_lock.path().to_string_lossy()
+            )
+        })?;
+
+        let v2 = (|| -> CargoResult<_> {
+            if unstable_upgrade {
+                let mut contents = String::new();
+                v2_lock
+                    .as_ref()
+                    .unwrap()
+                    .file()
+                    .read_to_string(&mut contents)?;
+                let mut v2 = if contents.is_empty() {
+                    CrateListingV2::default()
+                } else {
+                    serde_json::from_str(&contents)
+                        .chain_err(|| format_err!("invalid JSON found for metadata"))?
+                };
+                v2.sync_v1(&v1)?;
+                Ok(v2)
+            } else {
+                Ok(CrateListingV2::default())
+            }
+        })()
+        .chain_err(|| {
+            format_err!(
+                "failed to parse crate metadata at `{}`",
+                v2_lock.as_ref().unwrap().path().to_string_lossy()
+            )
+        })?;
+
+        Ok(InstallTracker {
+            v1,
+            v2,
+            v1_lock,
+            v2_lock,
+            unstable_upgrade,
+        })
+    }
+
+    /// Checks if the given package should be built, and checks if executables
+    /// already exist in the destination directory.
+    ///
+    /// Returns a tuple `(freshness, map)`. `freshness` indicates if the
+    /// package should be built (`Dirty`) or if it is already up-to-date
+    /// (`Fresh`). The map maps binary names to the PackageId that installed
+    /// it (which is None if not known).
+    ///
+    /// Returns an error if there is a duplicate and `--force` is not used.
+    pub fn check_upgrade(
+        &self,
+        dst: &Path,
+        pkg: &Package,
+        force: bool,
+        opts: &CompileOptions<'_>,
+        target: &str,
+        _rustc: &str,
+    ) -> CargoResult<(Freshness, BTreeMap<String, Option<PackageId>>)> {
+        let exes = exe_names(pkg, &opts.filter);
+        let duplicates = self.find_duplicates(dst, &exes);
+        if force || duplicates.is_empty() {
+            return Ok((Freshness::Dirty, duplicates));
+        }
+        // If any duplicates are not tracked, then --force is required.
+        // If any duplicates are from a package with a different name, --force is required.
+        let matching_duplicates: Vec<PackageId> = duplicates
```

Comment:
Is this a `Vec` to handle things like `foo v0.0.1` installs `foo` and `foo v0.0.2` installs `bar` and then `foo v0.0.3` installs both `foo` and `bar`?